controllers/metal3.io/hostfirmwaresettings_controller.go

```diff
@@ -0,0 +1,171 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"crypto/sha256"
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	metal3v1alpha1 "github.com/metal3-io/baremetal-operator/apis/metal3.io/v1alpha1"
+)
+
+// HostFirmwareSettingsReconciler reconciles a HostFirmwareSettings object
+type HostFirmwareSettingsReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+
+//+kubebuilder:rbac:groups=metal3.io,resources=hostfirmwaresettings,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=metal3.io,resources=hostfirmwaresettings/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=metal3.io,resources=firmwareschemas,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=metal3.io,resources=firmwareschemas/status,verbs=get;update;patch
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.6.4/pkg/reconcile
+func (r *HostFirmwareSettingsReconciler) Reconcile(ctx context.Context, req ctrl.Request) (result ctrl.Result, err error) {
+
+	reqLogger := r.Log.WithValues("hostfirmwaresettings", req.NamespacedName)
+	reqLogger.Info("start")
+
+	// Fetch the HostFirmwareSettings
+	hfs := &metal3v1alpha1.HostFirmwareSettings{}
+	if err = r.Get(ctx, req.NamespacedName, hfs); err != nil {
+		if k8serrors.IsNotFound(err) {
+			// Request object not found, could have been deleted after
+			// reconcile request. Return and don't requeue
+			reqLogger.Info("could not find host firmware settings")
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, errors.Wrap(err, "could not load host firmware settings")
+	}
+
+	// Check if settings has been cleared during pivot, and if so, restore firmwareSchema from annotation
+	if hfs.Status.FirmwareSchema == nil {
+		if _, ok := hfs.Annotations[metal3v1alpha1.FirmwareSchemaNameAnnotation]; ok {
+			hfs.Status.FirmwareSchema = &metal3v1alpha1.SchemaReference{
+				Name:      hfs.Annotations[metal3v1alpha1.FirmwareSchemaNameAnnotation],
+				Namespace: hfs.Annotations[metal3v1alpha1.FirmwareSchemaNamespaceAnnotation]}
+
+			if r.Status().Update(context.TODO(), hfs) != nil {
```

Comment:
So the purpose of this controller it's just to update the schema reference? Looks like a mixed responsibility situation between this controller and the baremetalhost one, which is slightly confusing for me.